### PR TITLE
[AArch64] Replace AND with LSL#2 for LDR target (#34101)

### PR DESCRIPTION
**Pull request by ParkHanbum**
Currently, process of replacing bitwise operations consisting of
`LSR`/`LSL` with `And` is performed by `DAGCombiner`.

However, in certain cases, the `AND` generated by this process
can be removed.

Consider following case:
```
        lsr x8, x8, #56
        and x8, x8, #0xfc
        ldr w0, [x2, x8]
        ret
```

In this case, we can remove the `AND` by changing the target of `LDR`
to `[X2, X8, LSL #2]` and right-shifting amount change to 56 to 58.

after changed:
```
        lsr x8, x8, #58
        ldr w0, [x2, x8, lsl #2]
        ret
```

This patch checks to see if the `SHIFTING` + `AND` operation on load
target can be optimized and optimizes it if it can.

### CODE REVIEW

**Comment by davemgreen** [comment]
Could this be handled better in DAG combine? Maybe with something similar to reassociationCanBreakAddressingModePattern?

**Comment by ParkHanbum** [comment]
ok. I'll find way

**Comment by davemgreen** [comment]
Hi. There is code in the DAG combiner to attempt to stop reassociations that would break the addressing mode.
https://github.com/llvm/llvm-project/blob/654846560c8892b32ea1e92c1fffc5e70bf0e75f/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2690
I've not looked into this case in a lot of details, there is a chance the transform happens in a few different places, but could the same thing be done here?

**Comment by ParkHanbum** [comment]
@davemgreen Thanks, I think I'm almost there thanks to your advice, I was thinking I should do it in `LOAD`. it's my mistake.

Can I ask one more thing?

I can check the legality of the address via `TLI.isLegalAddressingMode()`, but what function in the TLI can I utilize to check if `PreIndex` is allowed in architecture?

**Comment by ParkHanbum** [comment]
@davemgreen I think it works done. would you review again please?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
Comment (justinfargnoli) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:566]: What does `Lsl` stand for in the function name?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
Comment (justinfargnoli) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:9898]: Why not move the call to `isCanBeLoadedWithLsl()` outside of `SelectionDAG` and into the `aarch64` target?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
Comment (justinfargnoli) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:28368]: If `Load->getOpcode() == ISD::LOAD`, can `dyn_cast<LoadSDNode>(Load) == nullptr`?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
Comment (justinfargnoli) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:28359]: Nit: You could use [`dyn_cast_or_null`](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates).

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
Comment (justinfargnoli) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:28363]: Nit: If `Load->getOpcode() == ISD::LOAD` implies that `dyn_cast<LoadSDNode>(Load) == nullptr`, then you could use [`dyn_cast_or_null`](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates).

**Review by justinfargnoli — APPROVED** [review]
LGTM, but please wait for @davemgreen's review.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16924]: dyn_cast_or_null -> dyn_cast

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16925]: Can we change it from C2->getZExtValue() == 2 -> C2->getZExtValue() <= 3? That should capture the common cases where the shift is free.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16927]: I don't believe the _or_null is required. Can the `N->use_begin().getUse().getUser()` just be `*N->use_begin()`? It should also probably check that N->hasOneUse().

Same below for the load.

**Review by davemgreen — COMMENTED** [review]
Thanks, this does look like a good place for it.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16925]: Oh - should the shift amount match the loaded type? So if it is loading an i32 like the tests, it should be 2, but if it was loading i16/i64 it would need to be different shift amounts?

Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16925]: Yeah sounds good. A few other types sizes: https://godbolt.org/z/GqPT3KbbT. The shift in the addressing mode needs to match the loaded size.

Comment (ParkHanbum) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16925]: https://godbolt.org/z/d6xK3xvav

add comparing with ARM64-GCC.

Comment (ParkHanbum) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16925]: and it works done!

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16927]: You can probably remove the <= 3 now that ShlAmt is checked below.

Also it is probably worth adding a quick check that ShouldADD has Uses, as they are being accessed below.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (ParkHanbum) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16927]: did I properly?

Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16927]: I don't believe that use_begin will return null if there are no uses - it would try to access an invalid iterator. Maybe add a `&& ShouldADD->hasOneUse()` or `&& !ShouldADD->use_empty()` just to make sure it exists and we don't get any weird crashes.

**Comment by ParkHanbum** [comment]
I tested this with test-suite in m1 macbook and here is that results:

```
current main

Failed Tests (1):
  test-suite :: SingleSource/Regression/C/gcc-c-torture/execute/GCC-C-execute-alloca-1.test

```

```
This

Failed Tests (2):
  test-suite :: SingleSource/Regression/C/gcc-c-torture/execute/GCC-C-execute-alloca-1.test
  test-suite :: SingleSource/UnitTests/AArch64/aarch64-acle-fmv-features.test
```

what different result at test aarch64-acle-fmv-features.test between main and this. 

```
 22 bti
======main======
 23 simd
======this======
 23         FAIL
 24 simd
 25 fp
```

**Comment by davemgreen** [comment]
Does the test reliably fail with the new compiler? It sounds like it might be unrelated, as there have been a number of changes around FMV lately. The test looks like it fork()'s, so maybe something was going wrong with it? If it is a real failure cause by this patch then you might need to debug a bit to see what might be going wrong.

**Comment by ParkHanbum** [comment]
sorry, it was my mistake. I was tested it with my built llvm version. I have tested with `Apple clang version 15.0.0` and it passed.

**Comment by AtariDreams** [comment]
Any updates on this?

**Comment by ParkHanbum** [comment]
nope

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16930]: -> `(1ULL << ShlAmt) == ByteVT`

**Review by davemgreen — COMMENTED** [review]
Sorry - I think I believed you had commit access. I've given this an extra test and it seems OK. If you can update the suggestion to fix a warning, we can get this in.

**Review by davemgreen — APPROVED** [review]
Thanks (I didn't get notification for the update before)

**Comment by ParkHanbum** [comment]
@davemgreen never mind. I'm always thanks for your reviewing.

**Pull request merged**
